### ISSUE TITLE
Enhance no-multiple-empty-lines to allow different approaches for different block sizes

### ISSUE THREAD

**Issue by cbrwizard** [eslint/eslint#5734]
Hey, I have a use case to add a new option to the no-multiple-empty-lines rule.
I use Eslint 1.10.3, though latest (2.5.3) [docs](http://eslint.org/docs/rules/no-multiple-empty-lines) show that this option is not added yet to the rule.

Check out [this pull request](https://github.com/facebook/react/pull/6360) for React, click on 'Show 2 comments' on the outdated diff. The code was:

```js
describe('when event is `keypress`', function() {
  it('returns whatever getEventCharCode returns', function() {
    getEventCharCode.mockReturnValue(9001);
    var keyboardEvent = createEvent({type: 'keypress', charCode: 50});

    expect(keyboardEvent.which).toBe(9001);
  });
});

describe('when event is `keydown` or `keyup`', function() {
  it('returns a passed keyCode', function() {
    var keyboardEvent = createEvent({type: 'keyup', keyCode: 40});

    expect(keyboardEvent.which).toBe(40);
  });
});
```

Note that in every `describe` block there is an empty line before the line beginning with `expect`.
However, as it was noted by React team in that pull request, there is no need to add an empty line in the second `it` function, as it consists of just 2 not empty lines.

I wish there was a rule for that, which could give an error/warning when there is a not needed empty line in a small function.

Is it possible to add an option to a no-multiple-empty-lines rule for this use case? Like add an option 
```js
block-size-limits: [
  {minBlockSize: 2, max: 0},
  {minBlockSize: 3, max: 1},
  {minBlockSize: 10, max: 2},
]
```
I agree that this adds some complexity to the rule and might look scary. But it's a real use case I would like to have implemented. Have you got a better idea on how to approach this? If you're okay with it I can actually work on it [:

**Comment by mysticatea** [comment]
Thank you for this issue.

I don't think good that we make rules getting changeable setting by block size.
We have many rules related in line breaks (`padded-blocks`, `lines-around-comment`, `newline-after-var`, `newline-before-return`), so this would introduce much complexity.

I'd like to recommend to make a plugin for this.

**Comment by cbrwizard** [comment]
I agree, @mysticatea. This sounds like a good rule for a [eslint-plugin-jasmine](https://github.com/tlvince/eslint-plugin-jasmine).
However, I have another related idea. How about adding an option `minBlockSize` to a `newline-before-return` rule? I personally would like to use that rule, but I really don't want to force a new-line in a function like this (took from docs)
```js
function foo() {

    return;
}
```

**Comment by mysticatea** [comment]
@cbrwizard It's a bad case. :)

```js
function foo() {

    return;  /*error Unexpected newline before return statement.*/
}
```

**Comment by cbrwizard** [comment]
Oops @mysticatea you're right, that's a bad example [:
So basically, `newline-before-return` has a `minBlockSize`-ish option enabled by default and set to 1. How about giving an option to change it to a different number? Here is an example with this option set to 2.
Note: `minBlockSize` counts the lines with content inside the block, so it ignores empty lines inside. Probably a `minLOCInBlock` name is more proper but looks crazy.

The following patterns are considered a problem:

```js
/*eslint newline-before-return: ["error", {minBlockSize: 2}]*/

function foo() {
  var bar = 15;

  return bar;
}

function longFoo() {
  var bar = 15;
  console.log(bar);
  return bar;
}
```

The following patterns are not considered a problem:

```js
/*eslint newline-before-return: ["error", {minBlockSize: 2}]*/

function foo() {
  var bar = 15;
  return bar;
}

function longFoo() {
  var bar = 15;
  console.log(bar);

  return bar;
}
```

**Comment by mysticatea** [comment]
> So basically, newline-before-return has a minBlockSize-ish option enabled by default and set to 1.

No, it's checking whether a ReturnStatement exists at the head of a block or not.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
